File: Scala/src/com/yourtion/Pattern20/FlowControlExample.scala
package com.yourtion.Pattern20

/**
  * Created by Yourtion on 07/04/2017.
  */
object FlowControlExample {
  def choose[E](num: Int, first: () => E, second: () => E, third: () => E) =
    if (num == 1) first()
    else if (num == 2) second()
    else if (num == 3) third()

  def test[E](expression: E) = expression

  def testTwice[E](expression: E) = {
    expression
    expression
  }

  def printTwice[E](expression: E) {
    println(expression)
    println(expression)
  }

  def testByName[E](expression: => E) = {
    expression
    expression
  }

  def simplerChoose[E](num: Int, first: => E, second: => E, third: => E) =
    if (num == 1) first
    else if (num == 2) second
    else if (num == 3) third

  def timeRun[E](toTime: => E) = {
    val start = System.currentTimeMillis
    toTime
    System.currentTimeMillis - start
  }
  def avgTime[E](times: Int, toTime: => E) = {
    val allTimes = for (_ <- Range(0, times)) yield timeRun(toTime)
    allTimes.sum / times
  }

  def run(): Unit = {
    println("choose :")
    choose(2, () => println("hello, world"), () => println("goodbye, cruel world"), () => println("meh, indifferent world"))
    println("test :")
    test(println("hello, world"))
    println("testTwice :")
    testTwice(println("hello, world"))
    println("printTwice :")
    printTwice(5 * 5)
    println("testByName :")
    testByName(println("hello, world"))
    println("simplerChoose :")
    simplerChoose(2, println("hello, world"), println("goodbye, cruel world"), println("meh, indifferent world"))
    println()

    println("avgTime : ")
    println(avgTime(5, Thread.sleep(1000)))
  }
}

File: Clojure/src/com/yourtion/Pattern20/flow_control_example.clj
(ns com.yourtion.Pattern20.flow-control-example)

(defmacro make-logger [name form]
  `(defn ~name []
     (println (str "Calling " ~name))
     ~form))

(defn choose [num first second third]
  (cond
    (= 1 num) (first)
    (= 2 num) (second)
    (= 3 num) (third)))

(defmacro simpler-choose [num first second third]
  `(cond
     (= 1 ~num) ~first
     (= 2 ~num) ~second
     (= 3 ~num) ~third))

(defn time-run [to-time]
  (let [start (System/currentTimeMillis)]
    (to-time)
    (- (System/currentTimeMillis) start)))

(defmacro avg-time [times to-time]
  `(let [total-time#
         (apply + (for [_# (range ~times)] (time-run (fn [] ~to-time))))]
     (float (/ total-time# ~times))))

(defn avg-time-fn [times to-time]
  (let [total-time
        (apply + (for [_ (range times)] (time-run to-time)))]
    (float (/ total-time times))))

(let
  [total-time
   (apply + (for [_ (range 5)] (time-run (fn [] (Thread/sleep 100)))))]
  (float (/ total-time 5)))

(defn run []
  (println "choose :")
  (choose 2
          (fn [] (println "hello, world"))
          (fn [] (println "goodbye, cruel world"))
          (fn [] (println "meh, indifferent world")))
  (println "")

  (println "simpler-choose :")
  (simpler-choose 2
                  (println "hello, world")
                  (println "goodbye, cruel world")
                  (println "meh, indifferent world"))
  (println "")

  (println "avg-time :")
  (println (avg-time 5 (Thread/sleep 1000)))
  
  (println ""))
